### PR TITLE
Order execution

### PR DESCRIPTION
**Pull request by johnstonstott**
:--)

### CODE REVIEW

**Review by MaelSemler — APPROVED** [review]
Revolutiary stuff

File: warzone/Main.cpp [hunk 1]
```diff
@@ -3,9 +3,11 @@
...
 int main()
 {
-	return Cards::main()
+	/*return Cards::main()
 		| MapDriver::main()
 		| OrdersDriver::main()
 		| MapLoaderDriver::main()
-		| PlayerDriver::main();
+		| PlayerDriver::main();*/
+
```
Comment (jun-loke) [inline warzone/Main.cpp]: Just delete these. We aren't gonna use them

File: warzone/Map.cpp [hunk 2]
```diff
@@ -43,6 +52,19 @@ Continent* Territory::getContinent() {
 	return this->continent;
 }
 
+Player* Territory::getOwner() {
+	return this->owner;
+}
+
+void Territory::setOwner(Player* player) {
+	// TODO Do I have to dereference this or something?
+	this->owner = player;
```
Comment (jun-loke) [inline warzone/Map.cpp:61]: Looks fine to me

File: warzone/Orders.h [hunk 3]
```diff
@@ -14,112 +18,130 @@ class Order {
...
 public:
-	NegotiateOrder(Player* player);
+	NegotiateOrder(Player* player, Player* targeted);
 	NegotiateOrder(NegotiateOrder* other);
 	~NegotiateOrder();
 
 	bool validate();
 	bool execute();
-	string toString() const;
+	std::string toString() const;
```
Comment (jun-loke) [inline warzone/Orders.h:138]: Why does this have a toString? Shouldn't we just use <<?

File: warzone/OrdersDriver.cpp [hunk 4]
```diff
@@ -3,6 +3,18 @@
 
 
 int OrdersDriver::main() {
+	OrdersList* l1 = new OrdersList();
+	Deck* d1 = new Deck();
+	Player* p1 = new Player(nullptr, l1, d1);
+
+	// TODO
+	cout << "TODO" << endl;
```
Comment (jun-loke) [inline warzone/OrdersDriver.cpp:11]: LMAO

File: warzone/OrdersDriver.h [hunk 5]
```diff
@@ -5,4 +5,5 @@
 
 namespace OrdersDriver {
 	int main();
+	int oldMain();
```
Comment (jun-loke) [inline warzone/OrdersDriver.h]: Deleto

File: warzone/OrdersDriver.cpp [hunk 6]
```diff
@@ -3,6 +3,18 @@
...
+
+	return 0;
+}
+
+int OrdersDriver::oldMain() {
```
Comment (jun-loke) [inline warzone/OrdersDriver.cpp]: Feel free to deleto

File: warzone/Orders.h (hunk 3 above)
Comment (johnstonstott) [reply warzone/Orders.h:138]: Something about it being a subclass so it wasn't working? @felixlapierre helped me fix it but I'm not remembering the reason. It was like this for A1 also.

File: warzone/OrdersDriver.cpp (hunk 4 above)
Comment (johnstonstott) [reply warzone/OrdersDriver.cpp:11]: Worked really hard on this one

File: warzone/OrdersDriver.h (hunk 5 above)
Comment (johnstonstott) [reply warzone/OrdersDriver.h]: Just left it in so I could reference but will be sure to delete before final PR

File: warzone/Orders.cpp [hunk 7]
```diff
@@ -92,13 +83,15 @@ bool DeployOrder::execute() {
 		return false;
 	}
 
-	// TODO: Do actions once we have more details.
+	// Add the armies to the territory.
+	this->target->addArmies(this->numArmies);
 	cout << "Deploy order executed." << endl;
```
Comment (felixlapierre) [inline warzone/Orders.cpp:88]: Remove this print, we will use `toString` instead. Same for all orders.

File: warzone/Orders.cpp [hunk 8]
```diff
@@ -143,11 +140,55 @@ bool AdvanceOrder::execute() {
...
+			this->target->removeArmies(1); 
+		}
+		// Defender has 70% chance of killing attacker.
+		if ((1 + rand() % 100) < 70) {
+			this->source->removeArmies(1);
+		}
+	}
+
+	// When attacker wins, the survivors occupy the territory.
+	if (this->source->getArmies() > this->target->getArmies()) {
```
Comment (felixlapierre) [inline warzone/Orders.cpp:174]: what if they both end up with 0 armies? do we still transfer ownership to the attacker or does it remain owned by the defender?

File: warzone/Orders.cpp [hunk 9]
```diff
@@ -143,11 +140,55 @@ bool AdvanceOrder::execute() {
...
+
+		// Place remaining armies on the new territory.
+		this->source->removeArmies(numArmies);
+		this->target->addArmies(numArmies);
+
+		// Transfer ownership of territory to the winning player.
+		this->target->setOwner(this->source->getOwner());
+
+		// Add the new territory to the player's list of territories.
+		this->player->addTerritory(this->target);
```
Comment (felixlapierre) [inline warzone/Orders.cpp]: First you have to remove the territory from the defending player's list of territories.

File: warzone/Player.cpp [hunk 10]
```diff
@@ -69,6 +97,13 @@ void Player::setTerritories(vector<Territory*>* territoriesToAdd) //sets all the
 	this->playerTerritories = territoriesToAdd;
 }
 
+// Used to add a single territory to the list of territories.
+void Player::addTerritory(Territory* territoryToAdd)
+{
+	// TODO Do I need to dereference this?
+	this->playerTerritories->push_back(territoryToAdd);
```
Comment (felixlapierre) [inline warzone/Player.cpp]: No you do not need to dereference this

**Review by felixlapierre — COMMENTED** [review]
Most of the comments are minor, 1 of them might be a bug - i think you aren't removing a conquered territory from its previous owner's list of territories.

File: warzone/Orders.cpp (hunk 8 above)
Comment (johnstonstott) [reply warzone/Orders.cpp:174]: This was in the FAQ. The attacker only get the territory if it has > 0 surviving armies. So I believe this should cover that scenario.

File: warzone/Orders.cpp (hunk 7 above)
Comment (johnstonstott) [reply warzone/Orders.cpp:88]: I will remove these after writing my driver just to make it easier while I am testing it.

**Review by yabdelkawy — COMMENTED** [review]
Everything looks good to me! I believe Felix and Jun brought up some good points so once those are addressed, we're set to merge!

**Review by johnstonstott — COMMENTED** [review]
Good job

**Comment by johnstonstott** [comment]
> 
> 
> Good job

Thanks

**Comment by johnstonstott** [comment]
> 
> 
> > Good job
> 
> Thanks

Whoops wrong account

**Pull request merged**